ansible-galaxy - support resolvelib versions >= 0.5.3, < 0.9.0

s-hertel opened the pull request:
##### SUMMARY
Support recent resolvelib versions.

Related to #77624

Fixes https://github.com/ansible/ansible/issues/74569

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
ansible-galaxy - support resolvelib versions >= 0.5.3, <= 0.8.1

sivel commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -424,3 +454,64 @@ def get_dependencies(self, candidate):
+class CollectionDependencyProvider(CollectionDependencyProviderBase):
+
+    def _get_candidates_from_args(self, *args):
+        if SemanticVersion(resolvelib_version) < SemanticVersion("0.7.0"):
```
Just a thought here, what do you think about creating some "constants" at the module level that describe these version checks, to avoid the overhead of parsing and comparing multiple times?

Something like:
```python
RESOLVELIB_VERSION = SemanticVersion(resolvelib_version)
RESOLVELIB_060 = RESOLVELIB_VERSION  >= SemanticVersion("0.6.0")
RESOLVELIB_070 = RESOLVELIB_VERSION >= SemanticVersion("0.7.0")
RESOLVELIB_080 = RESOLVELIB_VERSION >= SemanticVersion("0.8.0")

[...]

if not RESOLVELIB_070:
[...]
```

Each one then becomes effectively: "at least x.y.z".  The overhead we experience is definitely more talking to APIs than this code, so maybe it's a premature optimization.  Just a thought I had when looking at this.

s-hertel replied:
Good idea, I updated this. I also refactored to create a provider class for each of the versions to keep it easily documentable.

s-hertel committed:
Reorganize code and fix download tests

Test incompatibilities are removed for resolvelib >= 0.6.0

Test against the latest 0.8.x version and fix requirements

s-hertel committed:
fix type hints

sivel commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -424,3 +478,63 @@ def get_dependencies(self, candidate):
+        return self._get_preference(list(candidates[identifier]))
+
+
+if RESOLVELIB_080:
+    CollectionDependencyProvider = CollectionDependencyProvider080  # type: ignore[assignment,no-redef,misc]
+elif RESOLVELIB_070:
+    CollectionDependencyProvider = CollectionDependencyProvider070  # type: ignore[assignment,no-redef,misc]
+elif RESOLVELIB_060:
+    CollectionDependencyProvider = CollectionDependencyProvider060  # type: ignore[assignment,no-redef,misc]
+else:
+    CollectionDependencyProvider = CollectionDependencyProvider050  # type: ignore[assignment,no-redef,misc]
```
I wonder if something like this would make mypy happier 🤷 

```suggestion
def _get_provider():  # type () -> CollectionDependencyProviderBase
    if RESOLVELIB_080:
        return CollectionDependencyProvider080
    elif RESOLVELIB_070:
        return CollectionDependencyProvider070
    elif RESOLVELIB_060:
        return CollectionDependencyProvider060
    return  CollectionDependencyProvider050


CollectionDependencyProvider = _get_provider()
```

s-hertel replied:
Yes, it seems to like that much better :smile:

s-hertel committed:
Improve mypy appeasement

mattclay commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -25,10 +25,17 @@
+from resolvelib import __version__ as resolvelib_version
+
+
+RESOLVELIB_VERSION = SemanticVersion.from_loose_version(LooseVersion(resolvelib_version))
+RESOLVELIB_060 = RESOLVELIB_VERSION >= SemanticVersion("0.6.0")
+RESOLVELIB_070 = RESOLVELIB_VERSION >= SemanticVersion("0.7.0")
+RESOLVELIB_080 = RESOLVELIB_VERSION >= SemanticVersion("0.8.0")
```
I'd move these 4 lines into `_get_provider` since it looks like they're not used anywhere outside that function.

Then the version range checks could just be inline with the `if` statements.

s-hertel replied:
Okay. The first is used by #77630, so I'm going to leave that one as-is, but the rest can just be removed as there's no need for the other constants.

s-hertel committed:
Remove extra stuff

mattclay commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -424,3 +475,66 @@ def get_dependencies(self, candidate):
+        return CollectionDependencyProvider080
+    elif RESOLVELIB_VERSION >= SemanticVersion("0.7.0"):
+        return CollectionDependencyProvider070
+    elif RESOLVELIB_VERSION >= SemanticVersion("0.6.0"):
```
These can be just `if` since the previous conditionals all use `return`.

s-hertel replied:
Done

s-hertel committed:
requested change

mattclay commented on requirements.txt:
```diff
@@ -10,4 +10,4 @@ packaging
 # NOTE: and we should update the upper cap with care, at least until 1.0
 # NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
-resolvelib >= 0.5.3, < 0.6.0  # dependency resolver used by ansible-galaxy
+resolvelib >= 0.5.3, < 0.9.0  # dependency resolver used by ansible-galaxy
```
We might want a note here to update the versions listed in the tests when updating this range.

s-hertel replied:
Good idea.

s-hertel committed:
add note

s-hertel committed:
Replace mypy | syntax with typing.Optional and typing.Union

sivel commented:
I'm really happy with the direction you've taken in this PR.  I like the implementation decision to separate out the classes.

I see expanded test coverage for the versions, so I think the only thing missing here is a changelog.

We'll need to discuss whether this should be backported or not.  I'm leaning toward the "no" side of the fence, but we can have a larger discussion once this is merged.

s-hertel committed:
Simplify find_matches

s-hertel commented:
Just for reference, these are the three commits that made changes to the find_matches/get_preference interfaces that this PR addresses:
https://github.com/sarugaku/resolvelib/commit/69f859b2e3564f278c235ff0c68bb85121df9721
https://github.com/sarugaku/resolvelib/commit/75e07e6b8cb7a3d5100bebdac02ebc359ec900c8
https://github.com/sarugaku/resolvelib/commit/02782f700b9d35698c5d5e54b7d9f4d199f88058

s-hertel committed:
changelog

s-hertel committed:
Fix tests - use a venv for testing the range of resolvelib versions

mattclay commented:
/azp run

s-hertel committed:
Remove vestigial line that should have been a merge conflict to be easily fixable during rebase

s-hertel committed:
Update temporary hardcoded fallback for ansible-test

s-hertel committed:
Update hardcoded upperbound for sanity tests

s-hertel committed:
Make error check more flexible

s-hertel commented:
/azp run

MarkusTeufelberger commented:
Can this be backported to the currently supported versions pretty please?

mattclay commented:
If this were to be backported, additional work would be required to update the test containers.

mattclay commented:
Ignore my previous comment. I was thinking the sanity requirements `.txt` files had been updated, but it's only the `.in` files, so it won't affect the sanity test environments.

s-hertel commented:
@MarkusTeufelberger I am not sure, but I opened #78008 for consideration.
